examples/NanoGPT/main.jl

```diff
@@ -0,0 +1,259 @@
+    text = replace(text, r"\r?\n" => " ")
+
+    ## an array of all unique characters
+    alphabet = [unique(text)..., '_']
+    stop = alphabet[end]
+
+    B = (length(text) - 1) ÷ sequence_length
+    # We must collect() before indexing, because String indexing does strange things with multi-byte
+    # characters and we could end up with the wrong length.
+    Xs = reshape(collect(text)[1:(B * sequence_length)], sequence_length, B)
+    Ys = reshape(collect(text)[2:(B * sequence_length + 1)], sequence_length, B)
+
+    # Input string starts with stop character '_', representing zero context.
+    Xs[1, :] .= stop
+
+    # Xs (input) should consist of indices into `alphabet` because this is what Embedding expects.
+    # Ys (output) should be one-hot because this is what logitcrossentropy expects.
+    Xs = map(c -> Int32(findfirst(==(c), alphabet)), Xs)
+    Ys = onehotbatch(Ys, alphabet)
+
+    trainX, testX = MLUtils.splitobs(Xs; at=1 - test_split)
+    trainY, testY = MLUtils.splitobs(Ys; at=1 - test_split)
+
+    return alphabet, Array(trainX), Array(trainY), Array(testX), Array(testY)
+end
+
+@main function main(;
+        n_embed::Int=64, n_hidden::Int=256, n_heads::Int=4, qk_dim::Int=16,
+        v_dim::Int=16, n_layers::Int=6, sequence_length::Int=64, batchsize::Int=128,
+        dropout_rate::Float32=0.0f0, test_split::Float64=0.1, lr::Float64=1e-2,
+        epochs::Int=100,
+        # Only inference options
+        inference::Bool=false, model_path::String="",
+        seed::Union{String, Vector{String}}=["_", "The", "Julia", "Lux.jl"],
+        output_length::Int=1024
+)
+    rng = Random.default_rng()
+    Random.seed!(rng, 1234)
+
+    dev = reactant_device()
+    cdev = cpu_device()
+
+    if inference
+        @printf "[Info] Inference mode enabled.\n"
+
+        @assert !isempty(model_path) "Please provide a path to a model checkpoint."
+
+        @printf "[Info] Loading model from %s.\n" model_path
+        model_config = JLD2.load(model_path, "model_config")
+        model = GPT(; model_config...)
+        ps = JLD2.load(model_path, "parameters")
+        st = JLD2.load(model_path, "states")
+        alphabet = JLD2.load(model_path, "alphabet")
+        sequence_length = model_config.sequence_length
+
+        texts = generate_text(
+            model, ps, st, seed; alphabet, output_length, sequence_length
+        )
+
+        for (i, (text, s)) in enumerate(zip(texts, seed))
+            @printf "[Info] Seed [%d]: %s\n" i s
+            @printf "[Generated Text] %s\n\n" text
+        end
+
+        return
+    end
+
+    alphabet, trainX, trainY, testX, testY = get_nanogpt_data(; sequence_length, test_split)
+
+    @printf "[Info] Alphabet size: %d\n" length(alphabet)
+    @printf "[Info] Training size: %d sequences.\n" size(trainX, 2)
+    @printf "[Info] Testing  size: %d sequences.\n\n" size(testX, 2)
+
+    train_loader = DataLoader(
+        (trainX, trainY); batchsize, shuffle=true, parallel=true
+    ) |> dev
```

Comment:
```suggestion
    ) .|> dev
```

Reply:
why is this needed? broadcasting over the dataloader will avoid constructing the DeviceIterator

Reply:
Atleast, on GPU without this the labels weren't being transferred to the device